### DISCUSSION TITLE
IR (TSOP4838) configuration

### DISCUSSION THREAD

**Discussion by alidaf** [sle118/squeezelite-esp32#298] [Q&A] [answered]
I have connected a TSOP4838 as follows:

Pin 1 -> GPIO34
Pin 2 -> GND
Pin 3 -> 3.3V

In NVS set_GPIO I have:

34=ir:nec

Is this correct?

I can't get it to work. I've swapped in a known working module from another project in case my soldering is to blame, tried a few different GPIO pins but nothing so far so I suspect it is this line in NVS.

**Comment by alidaf** [comment]
If it's any help, I'm measuring 3.3V on pin 1 as well as pin 3, which possibly indicates that it is connected correctly - the datasheet indicates that the output signal goes low at 250mV, but there's no response in LMS when a button is pressed. I don't have a scope so can't test. I'm using a Logitech Classic remote. LMS server settings is checked for slim_devices_remote and jvc_dvd.

**Comment by alidaf** [comment]
Ah, ok. If I take out the :nec, it works. Is this a bug?

**Comment by sle118** [comment]
I'll have to let @philippe44 answer this one

**Comment by philippe44** [comment]
I don't know... I can't reproduce and it seems that others can't either. Looking at the code, I don't see a potential issue and that's a few lines. I also don't quite understand your comment on the issues list https://github.com/sle118/squeezelite-esp32/issues/291#issuecomment-1716049183.
[edit]: oh, ok so you are 100000% sure that it works if you don't add :nec and that it does *not* work if you add it? Including making sure you rebooted between each test?

**Reply by alidaf** [reply, marked as answer]
Well, maybe 99999% sure! Definitely with reboots but I do have an insanely bad juju in my house. Just a change in wind direction can have the oddest of effects and remember the issues I was having with the rotary encoder button! I must have been evil in a previous life and I'm paying for it with bad karma. It all works now so I just want to get it all stuffed into the case and sort out the rear panel.

**Comment by Androssboy** [comment]
@philippe44 
Repeat is handled. I don't think I have an rc5 remote, all packets I receive are 34 in length which is NEC or Pronto (AFAIU). Probably your remote are not rc5 either. I've added a log where you can see the whole received buffer so you can figure out the coding and type of IR.

 It's a byte dump organized by 16 bits packets

 take 2 bytes, the LSByte is first so swap them

the 1st bit (MSbit) is the value read by the IR receiver, inverted
the next 15 bits are the duration in µs of the bit
and so on...
For example, using NEC protocol : bb 22 ee 91 is
bb 22 => 22 bb => first bit is 0 => 1, duration is 22 bb = 8891 µs
ee 91 => 91 ee => first bit is 1 => 0, duration is 11 ee = 4590 µs

This is a proper NEC header pattern. To analyze, chatGPT can be your friend. Something like "Take each segment in hexadecimal, assemble them by 2 to form a 16 bits value. Swap bytes. Take MSB, invert it, that's value. Take 15 LSB, that's duration in µs". And be careful, chatGPT spews a lot of crap as well.

> 

Dear philippe44 !
I'm amazed by your intelligence and skills! Thank you very much for your titanic work.
I'm trying to connect an IR remote control to an ESP32-A1S that has squeezelite-esp32 installed, but I can't succeed. I counted the remote control button codes using Arduino and did everything according to the instructions from the documentation on GIThub.
After carefully reading the forums, I found your message which I quoted at the beginning of this post.
Is this procedure for converting bits and bytes necessary for every remote control and every button? Or is it enough to read the button code using an Arduino sketch and paste it into Slim_Devices_Remote.ir on the LMS server? I'm completely confused...
Please forgive me, but I am NOT a programmer and I am trying to repeat the project with my own hands. By repeating after you, I myself become smarter!

**Comment by Androssboy** [comment]
> I have connected a TSOP4838 as follows:
> 
> Pin 1 -> GPIO34 Pin 2 -> GND Pin 3 -> 3.3V
> 
> In NVS set_GPIO I have:
> 
> 34=ir:nec
> 
> Is this correct?
> 
> I can't get it to work. I've swapped in a known working module from another project in case my soldering is to blame, tried a few different GPIO pins but nothing so far so I suspect it is this line in NVS.

@alidaf

Hello
Did you manage to solve the problem with IR control?
Please share your experience.
